forest/bonsai/simulate_gps_data.py

```diff
@@ -1117,31 +1122,261 @@ def process_switches(
     return switches
 
 
-def sim_GPS_data(cycle,p,data_folder):
-    ## only two parameters
-    ## cycle is the sum of on-cycle and off_cycle, unit is minute
-    ## p is the missing rate, in other words, the proportion of off_cycle, should be within [0,1]
-    ## it returns a pandas dataframe with only observations from on_cycles, which mimics the real data file
-    ## the data are the trajectories over two weeks, sampled at 1Hz, with an obvious activity pattern
-    s = datetime2stamp([2020,8,24,0,0,0],'America/New_York')*1000
-    if os.path.exists(data_folder)==False:
+def sim_gps_data(
+    N: int,
+    location: str,
+    start_date: List[int],
+    end_date: List[int],
+    cycle: int,
+    percentage: float,
+    api_key: str,
+    data_folder: str,
+    attributes_dir: Optional[str] = None,
+):
+    """Generates gps trajectories.
+
+    Args:
+        N: (int) number of people to simulate
+        location: (str) indicating country and city to simulate at,
+            format "Country_2_letter_ISO_code/City_Name"
+        start_date: (list) start date of trajectories,
+            format  [year, month, day]
+        end_date: (list) end date of trajectories,
+            end date is not included in the trajectories,
+            format [year, month, day]
+        cycle: (int) the sum of on-cycle and off_cycle,
+            unit is minute
+        percentage: (float) the missing rate, in other words,
+            the proportion of off_cycle, should be within [0,1]
+        api_key: (str), api key for open route service
+            https://openrouteservice.org/
+        data_folder: (str) directory to save trajectories
+        attributes_dir: (str) directory to json file
```

Comment:
It's a good practice to pass a file object instead of a file path. This allows to deal with IO errors externally and helps with testing. This would also help reduce the size of this function.